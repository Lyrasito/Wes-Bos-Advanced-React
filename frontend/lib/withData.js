import withApollo from "next-with-apollo";
import ApolloClient from "apollo-boost";
import { endpoint } from "../config";
import { LOCAL_STATE_QUERY } from "../components/Cart";

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === "development" ? endpoint : endpoint,
    request: (operation) => {
      operation.setContext({
        fetchOptions: {
          credentials: "include",
        },
        headers,
      });
    },
    //local data
    clientState: {
      resolvers: {
        //{cache} is destructured from client
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //read the cart state from the cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            //write cartOpen to the opposite
            const data = {
              data: {
                cartOpen: !cartOpen,
              },
            };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo(createClient);
