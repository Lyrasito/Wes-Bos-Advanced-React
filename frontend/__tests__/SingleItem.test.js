import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import SingleItemComponent, {
  SINGLE_ITEM_QUERY,
} from "../components/SingleItem";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeItem } from "../lib/testUtils";

describe("<SingleItemComponent />", () => {
  it("renders with proper data", async () => {
    const mocks = [
      {
        //when someone makes a request with this query and variable combo:
        request: { query: SINGLE_ITEM_QUERY, variables: { id: "123" } },
        //return this result:
        result: { data: { item: fakeItem() } },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItemComponent id="123" />
      </MockedProvider>
    );
    //initially, the call will be set at the "loading" phase because the data hasn't returned yet
    expect(wrapper.text()).toContain("Loading...");
    // wait() makes it so the next functions are called on the next tick, so after the data goes through. You also have to update the wrapper.
    await wait();
    wrapper.update();
    //don't snapshot the whole wrapper because the ApolloProvider will come along with it and it's loooong

    expect(toJSON(wrapper.find("h2"))).toMatchSnapshot();
    expect(toJSON(wrapper.find("img"))).toMatchSnapshot();
    expect(toJSON(wrapper.find("p"))).toMatchSnapshot();
  });
  it.only("errors with a not found item", async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: "123" } },
        //return this result, with errors inside:
        result: {
          errors: [{ message: "Item not found" }],
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItemComponent id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    //Wes put a prop of data-test to make it easier to find the elements we want in testing
    const item = wrapper.find("[data-test='graphql-error']");

    expect(item.text()).toContain("Item not found");
  });
});
