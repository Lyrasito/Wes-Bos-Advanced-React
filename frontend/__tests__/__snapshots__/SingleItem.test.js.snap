// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItemComponent /> renders with proper data 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`<SingleItemComponent /> renders with proper data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<SingleItemComponent /> renders with proper data 3`] = `
Array [
  <p>
    dogs
  </p>,
  <p>
    $50
  </p>,
]
`;
